lib/utils/fakeconn.go

```diff
@@ -201,6 +198,64 @@ func (c *ChConn) SetWriteDeadline(t time.Time) error {
 	return nil
 }
 
+// CancelableChConn is a wrapped SSH channel connection that supports deadlines.
+type CancelableChConn struct {
+	// ChConn is the base wrapped SSH channel connection.
+	*ChConn
+	// reader is the part of the pipe that clients read from.
+	reader net.Conn
+	// writer is the part of the pipe that receives data from SSH channel.
+	writer net.Conn
+}
+
+// NewCancelableChConn returns a new instance of wrapped SSH channel connection
+// that supports deadlines.
+func NewCancelableChConn(conn ssh.Conn, ch ssh.Channel) *CancelableChConn {
+	reader, writer := net.Pipe()
+	c := &CancelableChConn{
+		ChConn: NewChConn(conn, ch),
+		reader: reader,
+		writer: writer,
+	}
+	// Start copying from the SSH channel to the writer part of the pipe. The
+	// clients are reading from the reader part of the pipe (see Read below).
+	//
+	// This goroutine stops when either the SSH channel closes or this
+	// connection is closed e.g. by a http.Server (see Close below).
+	go io.Copy(writer, ch)
```

Comment:
nit: not sure if this goroutine races with `writer.Close` but I'd ensure that it exits before `CancelableChConn.Close` exits at least. On the bonus side, I'd ensure that it exits before the `writer` is closed,

Reply:
Yeah, I'm not sure there's necessarily a race - if any part of the connection (writer or ChConn) closes, the copy will just stop which isn't much different from just closing the connection during a read.

My primary concern with this was to avoid a goroutine leak which I think can't happen if any part of the connection closes (which it does). We could add a channel signaling that the copy has exited to be a bit paranoid though.